pkg/reconciler/configmappropagation/configmappropagation.go

```diff
@@ -0,0 +1,318 @@
+	})
+	current, err := r.configMapLister.ConfigMaps(cmp.Namespace).Get(expected.Name)
+	if err != nil && !apierrs.IsNotFound(err) {
+		logging.FromContext(ctx).Error("Unable to get ConfigMap: "+current.Name+" in current namespace", zap.Error(err))
+		return fmt.Errorf("error getting ConfigMap in current namespace: %v", err)
+	}
+
+	// Only update ConfigMap with knative.dev/eventing/config-propagation:copy label.
+	// If the ConfigMap does not have this label, the controller must not update the ConfigMap.
+	// Instead, it should record a k8s WARN-level Event on the ConfigMap.
+	if current != nil {
+		label := current.GetLabels()
+		if label[resources.PropagationLabelKey] != resources.PropagationLabelValueCopy {
+			return fmt.Errorf(`unable to update ConfigMap in current namespace, ConfigMap doesn't have "knative.dev/eventing/config-propagation:copy" label`)
+		}
+		if current, err = r.KubeClientSet.CoreV1().ConfigMaps(expected.Namespace).Update(expected); err != nil {
+			return fmt.Errorf("error updating ConfigMap in current namespace: %v", err)
+		}
+		return nil
+	}
+
+	if current, err = r.KubeClientSet.CoreV1().ConfigMaps(expected.Namespace).Create(expected); err != nil {
+		return fmt.Errorf("error creating ConfigMap in current namespace: %v", err)
+	}
+
+	return nil
+}
+
+// checkConfigMap will delete copy ConfigMap if original ConfigMap no longer exists or no longer has the required label
+func (r *Reconciler) checkConfigMap(ctx context.Context, cmp *v1alpha1.ConfigMapPropagation) error {
+	copyConfigMapList, err := r.configMapLister.ConfigMaps(cmp.Namespace).List(labels.Everything())
+	if err != nil {
+		logging.FromContext(ctx).Error("Unable to get the ConfigMap list in current namespace", zap.Error(err))
+		return err
+	}
+
+	var errs error
+	for _, copyConfigMap := range copyConfigMapList {
+		// Select copy ConfigMap which is controlled by current ConfigMapPropagation
+		if metav1.IsControlledBy(copyConfigMap, cmp) {
+			// Get the name of original ConfigMap
+			// The name of Copy ConfigMap is followed by <ConfigMapPropagation.Name>-<originalConfigMap.Name>
+			originalConfigMapName := strings.TrimPrefix(copyConfigMap.Name, cmp.Name+"-")
+			if err = r.deleteOrKeepConfigMap(ctx, cmp, copyConfigMap, originalConfigMapName); err != nil {
+				logging.FromContext(ctx).Warn("Failed to propagate ConfigMap: ", zap.Error(err))
+				r.Recorder.Eventf(cmp, corev1.EventTypeWarning, configMapPropagationPropagateSingleConfigMapFailed,
+					fmt.Sprintf("Failed to propagate ConfigMap %v: %v", originalConfigMapName, err))
+				if errs == nil {
+					errs = fmt.Errorf("one or more ConfigMap propagation failed")
+				}
+			}
+		}
+	}
+	return errs
+}
+
+func (r *Reconciler) deleteOrKeepConfigMap(ctx context.Context, cmp *v1alpha1.ConfigMapPropagation, copyConfigMap *corev1.ConfigMap, originalConfigMapName string) error {
+	originalConfigMap, err := r.configMapLister.ConfigMaps(cmp.Spec.OriginalNamespace).Get(originalConfigMapName)
```

Comment:
IIUC, this is kind of duplicated logic. In `reconcileConfigMap` you already list all the configmaps from the original namespace with propagation labels. So any configmap owned by the controller in the current namespace but not in the `originalConfigMapList` should be deleted. Is that right?

Reply:
This method is for the case if the `originalConfigMap` exists before but no longer here now, then the corresponding `copyConfigMap` should be deleted. According to my understanding, from `originalConfigMap` side, it can only propagate existing ConfigMap, but it can't know which `originalConfigMap` is deleted and propagate this already deleted `originalConfigMap`. So I check it from the `copyConfigMap` side. Do you think there is a better way to make that deletion?

Reply:
On L207 you select ConfigMaps from the original namespace to be `originalConfigMapList `. I think the logic for the current namespace could be simple as "if a managed ConfigMap is in the list, then update it; otherwise, delete it". Did I misunderstand something?